src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -117,18 +118,29 @@ private T runWithRetries(byte[] key, int redirections, boolean tryRandomNode, bo
       }
 
       return execute(connection);
-    } catch (JedisConnectionException jce) {
-      if (tryRandomNode) {
-        // maybe all connection is down
-        throw jce;
-      }
 
+    } catch (JedisNoReachableClusterNodeException jnrcne) {
+      //Not sure if JedisNoReachableClusterNodeException exception should extends JedisConnectionException, it's not connection exception
+      //Remove this if hierarchy reworked and no need to make exception from JedisConnectionException clause
+      throw jnrcne;
+    } catch (JedisConnectionException jce) {
       // release current connection before recursion
       releaseConnection(connection);
       connection = null;
 
-      // retry with random connection
-      return runWithRetries(key, redirections - 1, true, asking);
+      if (redirections <= 1) {
+        //We need this because if node is not reachable anymore - we need to finally initiate slots renewing,
+        //or we can stuck with cluster state without one node in opposite case.
+        //But now if maxRedirections = 1 or 2 we will do it too often. For each time-outed request.
```

Comment:
Yep, you will not experience it with maxRedirects=1, but for us it's bad solution and what's going on with non values > 1 looks strange)

Do you really want to get rid of cluster rediscovering on timeouts if you have maxRedirects=1?
I could implement solution with tracking success/failures and make rediscovering rarely than 1 per... some time like 5s and only if all requests failed. But it will be more complicated and required idea review firstly.